resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala

```diff
@@ -69,97 +72,191 @@ private[spark] class KerberosConfDriverFeatureStep(
-        hSpec.jobUserName,
-        krb5File,
-        Some(kubernetesConf.krbConfigMapName),
-        krb5CMap,
-        hadoopBasedSparkPod)
-    }.getOrElse(
-      HadoopBootstrapUtil.bootstrapSparkUserPod(
-        kubeTokenManager.getCurrentUser.getShortUserName,
-        hadoopBasedSparkPod))
   }
 
   override def getAdditionalPodSystemProperties(): Map[String, String] = {
-    val resolvedConfValues = kerberosConfSpec.map { hSpec =>
-      Map(KERBEROS_DT_SECRET_NAME -> hSpec.dtSecretName,
-        KERBEROS_DT_SECRET_KEY -> hSpec.dtSecretItemKey,
-        KERBEROS_SPARK_USER_NAME -> hSpec.jobUserName,
-        KRB5_CONFIG_MAP_NAME -> krb5CMap.getOrElse(kubernetesConf.krbConfigMapName))
-      }.getOrElse(
-        Map(KERBEROS_SPARK_USER_NAME ->
-          kubeTokenManager.getCurrentUser.getShortUserName))
-    Map(HADOOP_CONFIG_MAP_NAME ->
-      hadoopConfDirSpec.hadoopConfigMapName.getOrElse(
-      kubernetesConf.hadoopConfigMapName)) ++ resolvedConfValues
+    // If a submission-local keytab is provided, update the Spark config so that it knows the
+    // path of the keytab in the driver container.
+    if (needKeytabUpload) {
+      val ktName = new File(keytab.get).getName()
+      Map(KEYTAB.key -> s"$KERBEROS_KEYTAB_MOUNT_POINT/$ktName")
+    } else {
+      Map.empty
+    }
   }
 
   override def getAdditionalKubernetesResources(): Seq[HasMetadata] = {
-    val hadoopConfConfigMap = for {
-      hName <- newHadoopConfigMapName
-      hFiles <- hadoopConfigurationFiles
-    } yield {
-      HadoopBootstrapUtil.buildHadoopConfigMap(hName, hFiles)
+    Seq[HasMetadata]() ++ {
+      krb5File.map { path =>
+        val file = new File(path)
+        new ConfigMapBuilder()
+          .withNewMetadata()
+            .withName(kubernetesConf.krbConfigMapName)
+            .endMetadata()
+          .addToData(
+            Map(file.getName() -> Files.toString(file, StandardCharsets.UTF_8)).asJava)
+          .build()
+      }
+    } ++ {
+      // If a submission-local keytab is provided, stash it in a secret.
+      if (needKeytabUpload) {
+        val kt = new File(keytab.get)
+        Seq(new SecretBuilder()
+          .withNewMetadata()
+            .withName(ktSecretName)
+            .endMetadata()
+          .addToData(kt.getName(), Base64.encodeBase64String(Files.toByteArray(kt)))
+          .build())
+      } else {
+        Nil
+      }
+    } ++ {
+      if (delegationTokens != null) {
+        Seq(new SecretBuilder()
+          .withNewMetadata()
+            .withName(dtSecretName)
+            .endMetadata()
+          .addToData(KERBEROS_SECRET_KEY, Base64.encodeBase64String(delegationTokens))
+          .build())
+      } else {
+        Nil
+      }
     }
+  }
 
-    val krb5ConfigMap = krb5File.map { fileLocation =>
-      HadoopBootstrapUtil.buildkrb5ConfigMap(
-        kubernetesConf.krbConfigMapName,
-        fileLocation)
+  // Visible for testing.
+  def createDelegationTokens(): Array[Byte] = {
```

Comment:
nit: Why isn't this logic done in a `KubernetesDelegationTokenManager` or some Util class? It seems we are over-running this step with a lot of functionality? no? You are also just mocking via
`when(step).createDelegationToken()`, I personally don't agree with the coverage brought up by this method of unit-testing but this is an extension of the discussion on a separate PR. i.e. I would mock the tokenManager and test to ensure that the `obtainDelegationTokens` take in a particular `creds` as an argument (also, so that one could test the if statement below).`

Reply:
> Why isn't this logic done in a KubernetesDelegationTokenManager or some Util class?

Because that's not needed. What exactly would doing that achieve?

> I would mock the tokenManager and test to ensure that the obtainDelegationTokens take in a particular creds

That's not what this code is testing. This code is testing that when delegation tokens are created, they are stashed in a secret. For that I'm mocking the creation of delegation tokens, because without any services to talk to, delegation tokens will not be created.

(BTW, that's what mocks should be used for.)

Reply:
> I'm mocking the creation of delegation tokens

BTW there are other ways to achieve that, e.g. with `UserGroupInformation.createTestUser()` and `doAs`, but I think the current test is easier to understand.

Reply:
> Because that's not needed. What exactly would doing that achieve?

I mean that is the function of a DelegationTokenManager, no? To handle nuances related to DT management, both for creation and renewal. _shrug_

> This code is testing that when delegation tokens are created, they are stashed in a secret

Well there is logic being introduced here beyond just `tokenManager.obtainDelegationToken()`, right? And if the purpose of the tests is to take the results from delegation token creation and put into a secret, then it would make sense to mock a `KubernetesDelegtionTokenManager` and just call `kdtm.obtainTokens()`, where this logic would be housed. Else, if we were to keep this logic within this step, i'd prefer that all parts of the logic are mocked. 

Once again, a matter of opinion, that differs between unit-testing strategy. I'll loop in @mccheah

> I think the current test is easier to understand.

I agree, but is it comprehensive is what I am talking about here?

Reply:
> that is the function of a DelegationTokenManager. To handle nuances related to DT management

Yes. And the test here is not interested in whether DT creation works or not. It's interested in that if they are created, then they are stashed in a secret.

> then it would make sense to mock a KubernetesDelegtionTokenManager and just call kdtm.obtainTokens()

Would it make you happier if instead of having instead of the current method, have this:

```
def createDelegationTokens(creds: Credentials)
```

The only extra thing you'd really be testing is the call to `SparkHadoopUtil.get.serialize`. I don't really see a lot of gains there.

Reply:
> Hmm, I somewhat disagree, but maybe I am just overly opinionated.

The code that creates delegation tokens is not in this class. If delegation token creation is to be tested, it needs to be tested in the test code for `HadoopDelegationTokenManager`.

All this class does is take the tokens that other class creates and stash them in a secret.

I'll change this to not need a mock at all, mostly to make a point.